### ISSUE TITLE
[ENH] New API design for `pytorch-forecasting` v2

### ISSUE THREAD

**Issue by phoeenniixx** [jdb78/pytorch-forecasting#2407]
## Background
After some user feedback on the current API of `pytorch-forecasting`, we came to know there are some issues with it. 

### API 

- They hit many errors, and the messages did not say what went wrong or how to fix it
- Errors surfaced late, deep in the pipeline, rather than at the call that took the bad input
- Configs are plain dicts; users coming from `sklearn` and `torch` found this unintuitive
- Constructors accept invalid configs without complaint
- `fit` / `predict` do not check their inputs
- Users found the 3-stage pipeline (not using `pkg` class) easier to follow than the current entry point
- `predict` output and other returned objects are hard to inspect and do not behave as expected

### Docs

- Users could not find a copy-pasteable example of the v2 pipeline
- Docs do not describe the shape of the config dicts
- Users could not tell which parameters `model_cfg` accepts and got stuck there
- Online docs were hard to navigate and examples hard to find

## New v2 API

So, based on this input, we are revamping and making some (a LOT of) changes to the v2 API.
To keep the changes separate from the current version and to not break the current code, **we will be doing it on a separate branch (`v2-dev`) instead of `main`.**

### Changes

See #2398 for prototype.

- Make `TimeSeries` a datatype and not D1 layer - this will be the standardised datatype of `ptf-v2`. This means the input and final preds will be a `TimeSeries` obj.
- The metadata objects will now be dataclasses
- The `pkg` class will be renamed to `forecaster` (like in place of `TFT_pkg_v2`, it will be `TFTForecaster`)
  - The `model_cfg` will be replaced by the flattened params - meaning, the `forecaster` class will NOT accept the `model_cfg`, instead we will flatten the the cfg and the params will be passed directly to the class
  - It will now accept the datamodule and `trainer` class directly (not cfgs)
    - datamodule will be   passed as `__init__` argument
    - `trainer` can be passed as an `__init__` arg as well as as `fit` arg. 
      - Here, the `trainer` passed to `fit` will override the `trainer` passed to `__init__`.

### Vignettes
**Minimal Vignette**
```python
data = TimeSeries(
    df, 
    num=['x'], 
    ...,
) 
forecaster = TFTForecaster()
forecaster.fit(data)
```

**Maximal Vignette**
```python
trainer_init = Trainer(
    max_epochs = 5, 
    accelerator = "auto", 
    ...
)
trainer_fit = Trainer(
    max_epochs = 10, 
    accelerator = "cpu", 
    ...
)
data = TimeSeries(
    df, 
    num=['x'], 
    ...,
) 
forecaster = TFTForecaster(
    # model parameters
    n_layers=42,
    ...
    # data module parameters
    # this is a class that the user has written
    # (following a base pattern)
    datamodule = DecoderEncoderDataModuleCustomizedByUser(
        stuff=2,
        more_stuff=42,
    )
    trainer=trainer_init
)
forecaster.fit(data, trainer=trainer_fit) # trainer_fit overrides trainer_init
```

## Milestones

- [ ] Implement `TimeSeries` as datatype
- [ ] Update datamodule to accept data as `TimeSeries` object in place of a dataset obj
- [ ] Implement `forecaster` pipeline
- [ ] Update tests

**Comment by phoeenniixx** [comment]
FYI @CloseChoice @fkiraly

**Comment by fnhirwa** [comment]
## can we agree on the internal batch contract before the data-module work lands?

#2407 describes the public flow: `TimeSeries` in, a forecaster, and `TimeSeries` out. I think we also need to specify what the data modules pass to models so #2423, #2434, and the MultiLoss work can use the same contract.

### current state on `v2-dev` ( including PRs #2410, #2423)

| Boundary | Single target | Multiple targets |
|---|---|---|
| `TimeSeries.__getitem__` | `y: [T, 1]` | `y: [T, C]` |
| window `__getitem__` (both data modules) | `y: [H]` | `y: list[[H]]` |
| collate (both data modules) | `y: [B, H]` | `y: list[[B, H]]` |
| `MultiLoss.forward` expects | `(targets, sample_weight)` | `(targets, sample_weight)` |

here `B` is batch size, `H` is forecast horizon, and `C` is target count. `MultiLoss` instead expects `(targets, sample_weight)`. if it receives the bare target list, `y_actual[0][idx]` selects a batch row from the first target, while `y_actual[1]` is mistaken for a weight. #2436 adapts this in the model layer, but the list-versus-tensor branches remain. `PredictCallback` also indexes y positionally and cannot yet concatenate multi-target predictions returned as lists.

### batch contract i am proposing

one mapping returned by every v2 loader. Lightning and PyTorch already handle nested mappings for device transfer and pin-memory, so no custom hooks are needed.

```python
batch = {
    "inputs": {                  # named tensors, kept separate by dtype and rank
        "encoder_cat":  LongTensor[B, E, F_cat],
        "encoder_cont": FloatTensor[B, E, F_cont],
        "decoder_cat":  LongTensor[B, H, F_cat],
        "decoder_cont": FloatTensor[B, H, F_cont],
        "target_past":  FloatTensor[B, E, C],
        # lengths, masks, groups, model-specific fields
    },
    "target":        FloatTensor[B, H, C],   # C=1 keeps its axis; None only for unlabeled predict
    "sample_weight": FloatTensor[B, H] | None,
}
```

Rules:

- Data modules keep `[H, C]` per window and stack once to `[B, H, C]`. No `list | tuple` unions on `y` or `target_scale` anywhere in the training path.
- Target names and order come from validated `TimeSeries` / data-module metadata. Validate `C == len(target_names)` and that the loss covers exactly those names at `fit` time, which matches the "fail at the call that took the bad input" goal above.
- `inputs` stays a mapping. Packing `long` categoricals, `float` continuous, rank-2 static and rank-3 temporal features into one tensor forces casts and padding and hides the model interface.
- `target_scale` is left out of the contract until its meaning is unified: the encoder-decoder module computes a history magnitude, the tslib module carries a center/scale dict. Those are different objects.
- Scope `[B, H, C]` to numeric targets. Categorical targets would need a documented extension (likely a mapping of targets), not silent float coercion.

### prediction contract i am proposing

a single `[B, H, C]` output cannot hold a point head next to a quantile head. Key by target name for every target count:

```python
output = {
    "prediction": {
        "sales":   Tensor[B, H, 1],   # point
        "returns": Tensor[B, H, Q],   # quantile
    }
}
```

the metric adapter selects `batch["target"][..., i]` and `output["prediction"][name]` in metadata order and passes `(target_i, sample_weight)` to legacy metrics at that one boundary. `PredictCallback` concatenates per name. A named loss spec `{"sales": MAE(), "returns": QuantileLoss()}` would make ordering explicit; legacy `MultiLoss([...])` can be kept behind the same adapter.

### two unrelated meanings of "weight"

- `MultiLoss(metrics, weights=[...])`: one coefficient per target metric.
- `TimeSeries(weight="col")`: one value per time point, `[T]`. Neither data module currently carries it into the batch, and the torchmetrics wrapper rejects a non-`None` sample weight.

These must not be conflated. Decision needed: carry `sample_weight` as `[B, H]` in the first contract PR, or explicitly defer it.

### decisions requested

1. Batch is one mapping with `inputs` / `target` / `sample_weight`, no list/tuple unions.
2. `target` keeps its last axis for `C=1` (`[B, H, 1]`).
3. `inputs` stays a named mapping, not one packed tensor.
4. Predictions are a mapping keyed by target name for all `C`, with an adapter for models that return a tensor today.
5. `sample_weight`: support in the first PR, or defer.
6. `target_scale`: out of scope until encoder-decoder and tslib semantics are reconciled.

### suggested sequencing if agreed

1. Record the contract (with the three shape examples above) in this issue.
2. Data-module PR stacked on #2423: `[H, C]` through both window datasets, same mapping from both collate functions, remove the `list | tuple` checks.
3. Rebase #2436 as the model/metric PR consuming the mapping; tests assert the fixed format rather than list-vs-tensor.
4. Prediction/forecaster PR aligned with #2434: `predict_step`, `PredictCallback`, per-name concatenation, inverse transforms, output `TimeSeries` names.
5. Integration tests on both real data modules: `C=1` and `C>1`, mixed point/quantile widths, target coefficients, optional sample weights, `predict()` with returned names, mismatch errors.

happy to take items 2 and 3 if we decide to go this way:)
